### PR TITLE
sql,server: unify sqlInstanceDialer initialization using instancereader-based dialer.

### PR DESCRIPTION
**Pull request by shubhamdhama**
Previously, sqlInstanceDialer initialization paths differed based on the
environment: single-tenant environments used kvNodeDialer, which relies on
gossip layer information to resolve addresses, while multi-tenant
environments used instancereader.

This change unifies the initialization paths by using instancereader for
both environments, leveraging the availability of instance information even
in single-tenant setups.

Release note: None
Informs: https://github.com/cockroachdb/cockroach/issues/125336

### CODE REVIEW

File: pkg/rpc/nodedialer/nodedialer.go [hunk 1]
```diff
@@ -217,35 +218,36 @@ func (n *Dialer) ConnHealth(nodeID roachpb.NodeID, class rpc.ConnectionClass) er
...
 		return errors.New("no node dialer configured")
 	}
-	if err := n.rpcContext.ConnHealth(
-		addr, roachpb.NodeID(id), rpc.DefaultClass); err == nil {
-		return nil
+	addr, locality, err := n.resolver(roachpb.NodeID(id))
+	if err != nil {
+		return err
 	}
-	return n.rpcContext.GRPCDialPod(addr, id, roachpb.Locality{}, rpc.DefaultClass).Health()
```
Comment (shubhamdhama) [inline pkg/rpc/nodedialer/nodedialer.go:248]: I wonder if passing this empty locality had any implications earlier.

File: pkg/sql/distsql_physical_planner.go [hunk 2]
```diff
@@ -191,10 +189,10 @@ func NewDistSQLPlanner(
...
 		sqlInstanceDialer:    sqlInstanceDialer,
 		nodeHealth: distSQLNodeHealth{
-			gossip:             gw,
-			connHealthSystem:   connHealthCheckerSystem,
-			connHealthInstance: instanceConnHealthChecker,
-			isAvailable:        isAvailable,
+			gossip:                gw,
+			connHealth:            sqlInstanceDialer.ConnHealthTryDial,
+			connHealthWithResolve: sqlInstanceDialer.ConnHealthTryDialWithResolve,
+			isAvailable:           isAvailable,
```
Comment (shubhamdhama) [inline pkg/sql/distsql_physical_planner.go:195]: Maybe we can eliminate this whole structure or just eliminate `isAvailable`. And probably some other cleanups present in #127293

File: pkg/sql/distsql_physical_planner.go [hunk 3]
```diff
@@ -1269,7 +1267,7 @@ func (h *distSQLNodeHealth) checkSystem(
 		// artifact of rpcContext's reconnection mechanism at the time of
 		// writing). This is better than having it used in 100% of cases
 		// (until the liveness check below kicks in).
-		if err := h.connHealthSystem(roachpb.NodeID(sqlInstanceID), rpc.DefaultClass); err != nil {
+		if err := h.connHealthWithResolve(sqlInstanceID); err != nil {
```
Comment (shubhamdhama) [inline pkg/sql/distsql_physical_planner.go:1270]: As a follow-up we can explore if we can eleminate `checkSystem` function and just use `checkInstanceHealth` (then we wouldn't be needing `conHealthWithResolve`).

**Review by shubhamdhama — COMMENTED** [review]
Honestly, I'm not even sure if this is completely _correct_. But since the tests passed and based on my current intuition I think this could be right direction.

File: pkg/rpc/nodedialer/nodedialer.go (hunk 1 above)
Comment (andrewbaptist) [reply pkg/rpc/nodedialer/nodedialer.go:248]: The locality only matters for metrics. Ideally it should pass the `remoteLocality`, which is the Locality of the node it is talking to. I'm not sure if all the callers have the locality. If this connection is only opened to check health and then closed, it won't matter in any meaningful way, but is still better to fix.

File: pkg/sql/distsql_physical_planner.go (hunk 2 above)
Comment (andrewbaptist) [reply pkg/sql/distsql_physical_planner.go:195]: I need to pick up that PR again at some point (or someone could finish it). I can take a look again next week when I should have some time.

File: pkg/rpc/nodedialer/nodedialer.go (hunk 1 above)
Comment (shubhamdhama) [reply pkg/rpc/nodedialer/nodedialer.go:248]: Thanks for answering this. Yes, I've fixed it in this PR.

File: pkg/sql/distsql_physical_planner.go (hunk 2 above)
Comment (shubhamdhama) [reply pkg/sql/distsql_physical_planner.go:195]: If this PR looks correct, the next step is to use the work done in https://github.com/cockroachdb/cockroach/pull/128283 and see if we can use only `checkInstanceHealth` and remove `checkSystem`, thereby eliminating the usage of `kvNodeDialer`. If this all turns out well, we might only need a small subset of your PR  (I think just simplifying what `isAvailable` is doing here). I think I can see for your PR later since I'm in the context.

**Review by mgartner — COMMENTED** [review]
LGTM. I want to have someone else on queries take a quick look.

> Reviewed 5 of 5 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-:-O6bd5-hDRYMHMcgzIL-:bi3w5c7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @jeffswenson)

**Review by DrewKimball — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-:-O6cWwDO5rCikAgAnel1:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @jeffswenson, and @shubhamdhama)

___
*[`pkg/rpc/nodedialer/nodedialer.go` line 230 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-O6cVbd60qzWlZOqKYlV:-O6cVbd60qzWlZOqKYlW:bhfad27) ([raw file](https://github.com/cockroachdb/cockroach/blob/8da5c5e05f665dfc74a20db36807d9666d9455d6/pkg/rpc/nodedialer/nodedialer.go#L230)):*
> ```Go
> 	}
> 	if err := n.rpcContext.ConnHealth(
> 		addr, roachpb.NodeID(id), rpc.DefaultClass); err == nil {
> ```

I know we do it in a few other places, but is it safe to cast from `SqlInstanceID` to `NodeID`? A comment explaining why would be nice.
___
*[`pkg/rpc/nodedialer/nodedialer.go` line 240 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-O6cVy_-0GMKnWEq7msP:-O6cVy_-0GMKnWEq7msQ:b-9vi9kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8da5c5e05f665dfc74a20db36807d9666d9455d6/pkg/rpc/nodedialer/nodedialer.go#L240)):*
> ```Go
> // ConnHealthTryDialWithResolve returns nil if we have an open connection of the
> // request class to the given node that succeeded on its most recent heartbeat.
> // If no healthy connection is found, it will attempt to dial the node.
> ```

[nit] Why is the comment here corrected from `sqlinstance` to `node`, while the one on `ConnHealthTryDial` does the opposite, and both take `SqlInstanceID` as arguments?
___
*[`pkg/server/server_sql.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-O6cWCVh3ufMOeh-vMXE:-O6cWCVh3ufMOeh-vMXF:b3ghqi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7644eaf3bd1cc61a0ed36014be0c7297305add0a/pkg/server/server_sql.go#L613)):*
> ```Go
> 	// arranged for pod IDs and instance IDs to match since the
> 	// secondary tenant gRPC servers currently live on a different
> 	// port.
> ```

Looks like this comment might help explain my question above.

**Review by andrewbaptist — APPROVED** [review]
:lgtm:

Can you test to ensure this runs correctly if you are doing DistSQL planning across multiple SQL Pods. Otherwise this is a good step towards cleaning this up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-:-O7-Nai_CTxTG5WlM7Yh:bp4symh)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @jeffswenson and @shubhamdhama)

___
*[`pkg/rpc/nodedialer/nodedialer.go` line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130235#-O7-LmEv0hJg6EXGPsoO:-O7-LmEv0hJg6EXGPsoP:brlcueb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a8f9207a845b6d99d48c73c57e7f7a4ae94029c/pkg/rpc/nodedialer/nodedialer.go#L245)):*
> ```Go
> 	}
> 	if err := n.rpcContext.ConnHealth(
> 		addr, roachpb.NodeID(id), rpc.DefaultClass); err == nil {
> ```

Similar to the other comment, is this cast safe? I'm not sure how this would work in a separate process system where the SQL instances are not nodes. I know this was hear before, so maybe this is OK, but at least add a comment explaining why.
___
*[`pkg/sql/distsql_physical_planner.go` line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130235#gh-1746905428:-O7-MDWQ8C6XPle5FEgu:baaho2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/8da5c5e05f665dfc74a20db36807d9666d9455d6/pkg/sql/distsql_physical_planner.go#L195)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

If this PR looks correct, the next step is to use the work done in https://github.com/cockroachdb/cockroach/pull/128283 and see if we can use only `checkInstanceHealth` and remove `checkSystem`, thereby eliminating the usage of `kvNodeDialer`. If this all turns out well, we might only need a small subset of your PR  (I think just simplifying what `isAvailable` is doing here). I think I can see for your PR next since I'm in the context.
</blockquote></details>

Agreed. This PR makes sense, and is in the right direction.

**Comment by shubhamdhama** [comment]
**TL;DR:** This PR is blocked until we add support for `LocalityAddresses` for SQL instances.

The question above asks whether the following casting is correct:

```go
	addr, locality, err := n.resolver(roachpb.NodeID(id))
```

The answer is yes, based on the following definition of `addressResolver`:

```go
	addressResolver := func(nodeID roachpb.NodeID) (net.Addr, roachpb.Locality, error) {
		info, err := cfg.sqlInstanceReader.GetInstance(cfg.rpcContext.MasterCtx, base.SQLInstanceID(nodeID))
		if err != nil {
			return nil, roachpb.Locality{}, errors.Wrapf(err, "unable to look up descriptor for n%d", nodeID)
		}
		return &util.UnresolvedAddr{AddressField: info.InstanceRPCAddr}, info.Locality, nil
	}
```

In essence, the dialer converts from `SQLInstanceID` to `NodeID` and back to `SQLInstanceID` in the address resolver. In the single-tenant case, the NodeID is the same as the SQLInstanceID. Even in a multi-tenant scenario, where the `SQLInstanceID` may differ from the NodeID, this conversion back to the original ID ensures it still works.

However, the more significant question is whether this address resolver works for a single tenant in the same way as the previous `kvNodeDialer`. Upon investigation, we found that `kvNodeDialer` uses `LocalityAddress` (`--locality-advertise-addr`) and defaults to the `Address` (`--advertise-addr`) of the node descriptor. But in SQL instances, we only have the `Address` available. This could potentially degrade performance in certain cases.

[internal discussion](https://cockroachlabs.slack.com/archives/C02HWA24541/p1727085225479689)

**Comment by shubhamdhama** [comment]
Closing in favor of other work that will be done in #138715
